File: solutions/2026-07-07/javascript/2026-07-07-concatenate-non-zero-digits-and-multiply-by-sum-i.js
/**
 * @fileoverview LeetCode problem: Concatenate Non-Zero Digits and Multiply by Sum I
 * @problem_summary Given an integer n, form a new integer x by concatenating its non-zero digits. Calculate the sum of digits in x. Return x * sum.
 * @link https://leetcode.com/problems/concatenate-non-zero-digits-and-multiply-by-sum-i/
 * @approach
 * 1. Convert the input integer `n` to a string to easily iterate through its digits.
 * 2. Initialize `x_str` as an empty string to build the concatenated non-zero digits.
 * 3. Initialize `sum` to 0 to store the sum of non-zero digits.
 * 4. Iterate through each character (digit) of the string representation of `n`.
 * 5. If a digit is not '0':
 *    a. Append the digit to `x_str`.
 *    b. Convert the digit to an integer and add it to `sum`.
 * 6. If `x_str` is empty after iterating (meaning `n` had no non-zero digits or was 0), set `x` to 0. Otherwise, convert `x_str` to an integer for `x`.
 * 7. Return `x * sum`.
 * @time_complexity O(log10(n)). The time complexity is determined by the number of digits in n, which is proportional to the base-10 logarithm of n. Converting to string and iterating through digits takes logarithmic time.
 * @space_complexity O(log10(n)). The space complexity is determined by the string representation of `n` and the `x_str`, both of which have a length proportional to the number of digits in `n`.
 */

/**
 * @param {number} n
 * @return {number}
 */
const concatenateNonZeroAndMultiplyBySum = (n) => {
    // Convert the input integer to a string to easily access its digits.
    const nStr = n.toString();
    
    // Initialize an empty string to build the concatenated non-zero digits.
    let x_str = "";
    
    // Initialize a variable to store the sum of the non-zero digits.
    let sum = 0;

    // Iterate through each character (digit) of the string representation of n.
    for (let i = 0; i < nStr.length; i++) {
        const digitChar = nStr[i];
        
        // Check if the current digit is not '0'.
        if (digitChar !== '0') {
            // If it's a non-zero digit, append it to our concatenated string.
            x_str += digitChar;
            
            // Convert the digit character to an integer and add it to the sum.
            sum += parseInt(digitChar, 10);
        }
    }

    // Determine the value of x.
    // If x_str is empty (meaning n had no non-zero digits or was 0), x is 0.
    // Otherwise, convert the concatenated string of non-zero digits to an integer.
    const x = x_str === "" ? 0 : parseInt(x_str, 10);

    // Return the product of x and its sum of digits.
    return x * sum;
};
```

File: solutions/2026-07-07/python/2026-07-07-concatenate-non-zero-digits-and-multiply-by-sum-i.py
# Problem Summary:
# This problem requires forming a new integer 'x' by concatenating all non-zero digits of a given integer 'n' in their original order. If no non-zero digits are found, 'x' is 0. Then, calculate the sum of digits of 'x'. Finally, return the product of 'x' and its sum of digits.

# Link to the problem:
# https://leetcode.com/problems/concatenate-non-zero-digits-and-multiply-by-sum-i/

# Approach Explanation:
# 1. Convert the input integer 'n' to its string representation to easily iterate through its digits.
# 2. Iterate through the characters of the string. Collect all characters that are not '0' into a list.
# 3. Join these non-zero digit characters to form a new string, let's call it `x_str`.
# 4. If `x_str` is empty (meaning 'n' had no non-zero digits, e.g., n=0), then 'x' is 0. Otherwise, convert `x_str` to an integer to get `x_val`.
# 5. Calculate the sum of digits of `x_val`. This can be done by repeatedly taking `x_val % 10` to get the last digit and `x_val //= 10` to remove it, until `x_val` becomes 0.
# 6. Finally, return the product of `x_val` and the calculated sum of its digits.

# Time Complexity Analysis:
# The conversion of 'n' to a string takes O(log10(n)) time, as the number of digits is proportional to log10(n).
# Iterating through the string and filtering non-zero digits takes O(log10(n)) time.
# Joining the characters and converting `x_str` to an integer `x_val` takes O(log10(x)) time, which is at most O(log10(n)).
# Calculating the sum of digits of `x_val` also takes O(log10(x)) time, which is at most O(log10(n)).
# Therefore, the dominant operation is proportional to the number of digits in 'n'.
# Overall Time Complexity: O(log10(n)).

# Space Complexity Analysis:
# Storing the string representation of 'n' takes O(log10(n)) space.
# Storing the list of non-zero digits and the string `x_str` also takes O(log10(n)) space in the worst case (e.g., n=123456789).
# Other variables like `x_val`, `digit_sum`, `current_x` use O(1) space.
# Overall Space Complexity: O(log10(n)).
# Given n <= 10^9, log10(n) is at most 10, so this is very efficient in practice.

class Solution:
    def concatenateNonZeroDigits(self, n: int) -> int:
        # Convert the integer n to a string to easily process its digits.
        n_str = str(n)
        
        # Initialize a list to store non-zero digits as strings.
        non_zero_digits_list = []
        
        # Iterate through each character (digit) in the string representation of n.
        for digit_char in n_str:
            # If the digit is not '0', add it to our list.
            if digit_char != '0':
                non_zero_digits_list.append(digit_char)
        
        # Join the collected non-zero digits to form the string representation of x.
        x_str = "".join(non_zero_digits_list)
        
        # Convert x_str to an integer x_val.
        # If x_str is empty (meaning n had no non-zero digits, e.g., n=0), x_val should be 0.
        x_val = int(x_str) if x_str else 0
        
        # Calculate the sum of digits of x_val.
        current_x = x_val
        sum_of_digits_x = 0
        
        # Loop until current_x becomes 0.
        # If x_val is 0, the loop will not execute, and sum_of_digits_x will remain 0, which is correct.
        while current_x > 0:
            # Add the last digit of current_x to sum_of_digits_x.
            sum_of_digits_x += current_x % 10
            # Remove the last digit from current_x by integer division.
            current_x //= 10
            
        # Return the product of x_val and its sum of digits.
        return x_val * sum_of_digits_x